Feat: add RIP-7212 to the list of precompiles Cairo1Helpers module

Quentash opened the pull request:
<!--- Please provide a general summary of your changes in the title above -->
Add the precompiled contracts p256verify to the list of precompiled contracts at address 0x0b following this implementation : https://github.com/ethereum/go-ethereum/pull/27540 .

I have chosen the 0x0b instead of 0x100 because of the amount of lines that the second one would have added to the precompiles.cairo and I am currently unaware of a precompiled contract at address 0x0b.

## Pull Request type

<!-- Please try to limit your pull request to one type; submit multiple
pull requests if needed. -->

Please check the type of change your PR introduces:

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying,
or link to a relevant issue. -->

Resolves: #755 

## What is the new behavior?

<!-- Please describe the behavior or changes that are being added by this PR. -->

- p256verify is now added to the list of precompiled contract at address 0x0b

## Does this introduce a breaking change?

- [ ] Yes
- [X] No

<!-- If this does introduce a breaking change, please describe the impact and
migration path for existing applications below. -->

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/779)
<!-- Reviewable:end -->

enitrat commented:
> I have chosen the 0x0b instead of 0x100 because of the amount of lines that the second one would have added to the precompiles.cairo and I am currently unaware of a precompiled contract at address 0x0b.

You can just add a specific if / else before the match table handling this specific case for 0x100. This should not be an issue when deciding a precompile address. Given that it seems a consensus was found on using 0x100 - 0x1ff for rollup precompiles, let's go with 0x100

enitrat reviewed (changes requested):
> Reviewed 4 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/779)* status: 4 of 5 files reviewed, 3 unresolved discussions (waiting on @ClementWalter and @Quentash)

___
*[`crates/evm/src/precompiles.cairo` line 72 at r1](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/779#-Nvlby87Dlj5WKxUVsz7:-Nvlby87Dlj5WKxUVsz8:b-11knjb) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/16fc6f633b97fa82a57b192eed1ee605c33f3695/crates/evm/src/precompiles.cairo#L72)):*
> ```Cairo
>                 )
>             },
>             11 => { P256Verify::exec(input)? },
> ```

see main comment - we can still move use it as 0x100
___
*[`crates/evm/src/precompiles/p256verify.cairo` line 20 at r1](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/779#-NvlcnFU6uyzeyUkv6j5:-NvlcnFU6uyzeyUkv6j6:b-ipv4a1) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/16fc6f633b97fa82a57b192eed1ee605c33f3695/crates/evm/src/precompiles/p256verify.cairo#L20)):*
> ```Cairo
>     fn exec(input: Span<u8>) -> Result<(u128, Span<u8>), EVMError> {
>         let gas: u128 = P256VERIFY_PRECOMPILE_GAS_COST;
> ```

missing input size check. if size is 160 then you can assume the precompile input is correct and simply `message_hash.from_be_bytes().unwrap()` (same for all conversions)

globally I think you can remove all the manual error handling as long as size is ok. if size !=160 return 0
___
*[`crates/evm/src/tests/test_precompiles/test_p256verify.cairo` line 16 at r1](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/779#-NvlddS95FviOt68DNVI:-NvlddS95FviOt68DNVJ:bnll29b) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/16fc6f633b97fa82a57b192eed1ee605c33f3695/crates/evm/src/tests/test_precompiles/test_p256verify.cairo#L16)):*
> ```Cairo
> #[test]
> fn test_p256verify_precompile() {
>     let (_, _) = setup_contracts_for_testing();
> ```

do you need to setup a contract? Can this not be a unit test?


<!-- Sent from Reviewable.io -->

Quentash commented:
- Chose 0x100 as precompiled address
- Added length check ( nice catch )
- Added tests concerning length check
- Removed setup_contracts for unit tests

enitrat reviewed (changes requested):
> Reviewed 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/779)* status: all files reviewed, 3 unresolved discussions (waiting on @ClementWalter and @Quentash)

___
*[`crates/evm/src/call_helpers.cairo` line 124 at r2](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/779#-Nvls41sD-vK91EmBmvv:-Nvls41sD-vK91EmBmvw:b-rnf1pp) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/759c8f76dd4f239e02c2242774cd5028922fe284/crates/evm/src/call_helpers.cairo#L124)):*
> ```Cairo
> fn is_precompile(self: EthAddress) -> bool {
>     let self: felt252 = self.into();
>     return (self != 0 && self.into() < 257_u256);
> ```

only [0, 9] and 0x100 are valid addresses, this needs to be modified accordingly


<!-- Sent from Reviewable.io -->

Quentash commented:
Corrected precompiled address range

enitrat commented:
some tests are failing, can you verify is `256` or `0x100` is an address used in tests?

Quentash commented:
Alright, my bad, I missed those since I struggle running every tests and didn't notice it on github.
But indeed, some tests were deploying on the 0x100 address which created conflicts. I changed them to deploy their test contract on 0x101.

enitrat commented:
> Alright, my bad, I missed those since I struggle running every tests and didn't notice it on github.
> But indeed, some tests were deploying on the 0x100 address which created conflicts. I changed them to deploy their test contract on 0x101.

0x101 is still too "close" we should ideally have a random value like 0xabfa740ccd

Quentash commented:
You're right ! 0xabfa740ccd it is then !

enitrat reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r3, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/779#-:-Nvplc8805D4NDB_IRhS:bnfp4nl)* status: all files reviewed, 3 unresolved discussions (waiting on @ClementWalter and @Quentash)



<!-- Sent from Reviewable.io -->
